lib/puppet/file_system/windows.rb

```diff
@@ -89,6 +99,25 @@ def read_preserve_line_endings(path)
 
   private
 
+  def expand_path_cached(path)
+    @memoized_expanded_paths ||= {}
```

Comment:
This only tracks dirs with '~' in them, and maintains a string to string map like:

* `c:\users\admini~1` -> `c:\users\Administrator`

I don't believe any of your concerns should apply.

I will probably end up removing this bit though because I think it adds only a negligible perf gain, but additional code / complexity.